v5/core/cr_authenticator.go

```diff
@@ -0,0 +1,603 @@
+		DisableSSLVerification:     disableSSLVerification,
+		Scope:                      scope,
+		Headers:                    headers,
+	}
+
+	// Make sure the config is valid.
+	err := authenticator.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	return authenticator, nil
+}
+
+// newComputeResourceAuthenticatorFromMap constructs a new ComputeResourceAuthenticator instance from a map containing
+// configuration properties.
+func newComputeResourceAuthenticatorFromMap(properties map[string]string) (authenticator *ComputeResourceAuthenticator, err error) {
+	if properties == nil {
+		return nil, fmt.Errorf(ERRORMSG_PROPS_MAP_NIL)
+	}
+
+	// Grab the AUTH_DISABLE_SSL string property and convert to a boolean.
+	disableSSL, err := strconv.ParseBool(properties[PROPNAME_AUTH_DISABLE_SSL])
+	if err != nil {
+		disableSSL = false
+	}
+
+	authenticator, err = NewComputeResourceAuthenticator(
+		properties[PROPNAME_CRTOKEN_FILENAME],
+		properties[PROPNAME_INSTANCE_METADATA_SERVICE_URL],
+		properties[PROPNAME_IAM_PROFILE_NAME],
+		properties[PROPNAME_IAM_PROFILE_ID],
+		properties[PROPNAME_AUTH_URL],
+		properties[PROPNAME_CLIENT_ID],
+		properties[PROPNAME_CLIENT_SECRET],
+		disableSSL,
+		properties[PROPNAME_SCOPE],
+		nil)
+
+	return
+}
+
+// AuthenticationType returns the authentication type for this authenticator.
+func (*ComputeResourceAuthenticator) AuthenticationType() string {
+	return AUTHTYPE_CRAUTH
+}
+
+// Authenticate adds IAM authentication information to the request.
+//
+// The IAM access token will be added to the request's headers in the form:
+//
+// 		Authorization: Bearer <access-token>
+//
+func (authenticator *ComputeResourceAuthenticator) Authenticate(request *http.Request) error {
+	token, err := authenticator.GetToken()
+	if err != nil {
+		return err
+	}
+
+	request.Header.Set("Authorization", "Bearer "+token)
+	return nil
+}
+
+// getTokenData returns the tokenData field from the authenticator with synchronization.
+func (authenticator *ComputeResourceAuthenticator) getTokenData() *iamTokenData {
+	authenticator.tokenDataMutex.Lock()
+	defer authenticator.tokenDataMutex.Unlock()
+
+	return authenticator.tokenData
+}
+
+// setTokenData sets the 'tokenData' field in the authenticator with synchronization.
+func (authenticator *ComputeResourceAuthenticator) setTokenData(tokenData *iamTokenData) {
+	authenticator.tokenDataMutex.Lock()
+	defer authenticator.tokenDataMutex.Unlock()
+
+	authenticator.tokenData = tokenData
+}
+
+// Validate the authenticator's configuration.
+//
+// Ensures that one of IAMProfileName or IAMProfileID are specified, and the ClientId and ClientSecret pair are
+// mutually inclusive.
+func (authenticator *ComputeResourceAuthenticator) Validate() error {
+
+	// Check to make sure that one of IAMProfileName or IAMProfileID are specified.
+	if authenticator.IAMProfileName == "" && authenticator.IAMProfileID == "" {
+		return fmt.Errorf(ERRORMSG_EXCLUSIVE_PROPS_ERROR, "IAMProfileName", "IAMProfileID")
```

Comment:
@padamstx I was reading through this while working on my implementation and I noticed a small detail that I didn't pick up on in my review - I don't think this is the right error message for this situation. The IAM Profile name and ID can _both_ be specified, right? But this error message says that exactly one must be specified

Reply:
yes, you're right... As I recall, I consciously used this pre-existing error message (rather than create a new one) for the case where neither of the properties are specified.   Technically it's not 100% correct because, as you say, they can in fact be specified together (as long as they are consistent with each other).
I can open an issue to fix this by introducing a new error message that covers this case.